File: 20240314_Thursday/20240314.py
# https://leetcode.com/problems/binary-subarrays-with-sum/solution/
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        total_count = 0
        current_sum = 0
        # {prefix: number of occurrence}
        freq = {}  # To store the frequency of prefix sums

        for num in nums:
            current_sum += num
            if current_sum == goal:
                total_count += 1

            # Check if there is any prefix sum that can be subtracted from the current sum to get the desired goal
            if current_sum - goal in freq:
                total_count += freq[current_sum - goal]

            freq[current_sum] = freq.get(current_sum, 0) + 1

        return total_count

File: 20240314_Thursday/20240314.go
// Leetcode: 930. Binary Subarrays With Sum.
// https://leetcode.com/problems/binary-subarrays-with-sum/

package main

import "fmt"
var p = fmt.Println

func numSubarraysWithSum(nums []int, goal int) int {
	// curr_sum - prev_sum = goal
	// curr_sum - goal = prav_sum
	// Take look prev_sum
	h := make(map[int]int)
	h[0] = 1
	ans := 0
	currSum := 0
	for i, v := range nums {
		currSum += v
		prevSum := goal - currSum
		p(i,currSum, prevSum, h[prevSum])
		ans += h[prevSum]
		h[currSum] += 1
	}
	return ans
}
